[JSC] Support ArrayWithInt32.concat(ArrayWithDouble) fast path

Constellation opened the pull request:
#### 0ac3e5f3c4d0c91cfc3548c1a41b82cbb8735707
<pre>
[JSC] Support ArrayWithInt32.concat(ArrayWithDouble) fast path
<a href="https://bugs.webkit.org/show_bug.cgi?id=258340">https://bugs.webkit.org/show_bug.cgi?id=258340</a>
rdar://111085265

Reviewed by Michael Saboff.

This patch adds ArrayWithInt32.concat(ArrayWithDouble) fast path by promoting from Int32 to Double array.

                                              ToT                     Patched

    array-concat-int32-and-double        1.6732+-0.0203     ^      1.2596+-0.0190        ^ definitely 1.3283x faster
    array-concat-double-and-int32        1.5538+-0.0151     ^      1.2633+-0.0136        ^ definitely 1.2299x faster

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::concatAppendOne):
(JSC::copyElements):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::appendMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::JSArray::mergeIndexingTypeForCopying):

Canonical link: <a href="https://commits.webkit.org/265386@main">https://commits.webkit.org/265386@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fc07f851fefb30de9be2662938bd83d7b8576f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11285 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12414 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10323 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10961 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13228 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10925 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11839 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/9053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12818 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9133 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9711 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16969 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9120 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13114 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10215 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10336 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8426 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9494 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2956 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13767 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11183 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1211 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10197 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2731 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1471,7 +1471,8 @@ static JSValue concatAppendOne(JSGlobalObject* globalObject, VM& vm, JSArray* fi
 
     unsigned resultSize = checkedResultSize;
-    IndexingType type = first->mergeIndexingTypeForCopying(indexingTypeForValue(second) | IsArray);
+    bool allowPromotion = false;
```
Use enum?

Constellation replied:
I think that bool is enough here. The reason is that these enums are used to make the code readable and we explicitly declares the passing bool with variable name to make it clearly readable here. (That is avoiding just passing true/false)

hyjorc1 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1566,7 +1587,8 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoPrivateFuncConcatMemcpy, (JSGlobalObject* glo
     IndexingType firstType = firstArray->indexingType();
     IndexingType secondType = secondArray->indexingType();
-    IndexingType type = firstArray->mergeIndexingTypeForCopying(secondType);
+    bool allowPromotion = true;
```
Ditto

Constellation replied:
Ditto

hyjorc1 commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -488,7 +488,8 @@ bool JSArray::appendMemcpy(JSGlobalObject* globalObject, VM& vm, unsigned startI
     IndexingType type = indexingType();
     IndexingType otherType = otherArray->indexingType();
-    IndexingType copyType = mergeIndexingTypeForCopying(otherType);
+    bool allowPromotion = false;
```
Ditto

Constellation replied:
Ditto

hyjorc1 commented on Source/JavaScriptCore/runtime/JSArray.h:
```diff
@@ -111,7 +111,7 @@ class JSArray : public JSNonFinalObject {
     bool canDoFastIndexedAccess() const;
     // This function returns NonArray if the indexing types are not compatable for copying.
-    IndexingType mergeIndexingTypeForCopying(IndexingType other);
+    IndexingType mergeIndexingTypeForCopying(IndexingType other, bool allowPromotion);
```
Ditto

Constellation replied:
Ditto

hyjorc1 commented on Source/JavaScriptCore/runtime/JSArrayInlines.h:
```diff
@@ -31,7 +31,7 @@
 namespace JSC {
 
-inline IndexingType JSArray::mergeIndexingTypeForCopying(IndexingType other)
+inline IndexingType JSArray::mergeIndexingTypeForCopying(IndexingType other, bool allowPromotion)
```
Ditto

Constellation replied:
Ditto

hyjorc1 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1511,16 +1512,36 @@ void clearElement(double& element)
-    for (unsigned i = sourceSize; i--;)
-        clearElement<T>(buffer[i + offset]);
+    if constexpr (std::is_same_v<T, U>) {
+        gcSafeMemcpy(buffer + offset, source, sizeof(JSValue) * sourceSize);
```
Why we use `sizeof(JSValue)` instead of `sizeof(U)` for the size of source element type?

Constellation replied:
They are always the same size to sizeof(JSValue). This function is specifically for butterfly’s buffer filling. So sizeof(JSValue) is rather better.

msaboff commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -1511,16 +1512,36 @@ void clearElement(double& element)
-    if (sourceType != ArrayWithUndecided) {
-        gcSafeMemcpy(buffer + offset, source, sizeof(JSValue) * sourceSize);
+    if (sourceType == ArrayWithUndecided) {
+        for (unsigned i = sourceSize; i--;)
```
Any reason to copy backwards, beside copying the old code?

Constellation replied:
I think there is no particular reason. We should just make it forward copy, changed.

msaboff reviewed (approved):
r=me with question

Constellation committed:
[JSC] Support ArrayWithInt32.concat(ArrayWithDouble) fast path
https://bugs.webkit.org/show_bug.cgi?id=258340
rdar://111085265

Reviewed by Michael Saboff.

This patch adds ArrayWithInt32.concat(ArrayWithDouble) fast path by promoting from Int32 to Double array.

                                              ToT                     Patched

    array-concat-int32-and-double        1.6732+-0.0203     ^      1.2596+-0.0190        ^ definitely 1.3283x faster
    array-concat-double-and-int32        1.5538+-0.0151     ^      1.2633+-0.0136        ^ definitely 1.2299x faster

* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::concatAppendOne):
(JSC::copyElements):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::appendMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::JSArray::mergeIndexingTypeForCopying):

Canonical link: https://commits.webkit.org/265386@main
